### PR TITLE
K8SPXC-442: Add finalizer for S3 backups

### PR DESCRIPTION
**Pull request by blez**
[K8SPXC-442](https://jira.percona.com/browse/K8SPXC-442) [&#10088;?&#10089;](https://pullrequestbadge.com/?utm_medium=github&utm_source=percona&utm_campaign=badge_info)

This PR provides possibility to add finalizers for backups on S3.
If user deletes backup and finalizer exists reconcile loop will start to delete files from S3
Deletion process is done as a background operation (run as separate gorutine) to not block whole reconcile loop during the time while backup deletion is in progress. But to not spin up an unpredictable count of gorutines we have a limit (default is 10) which can be configured through env var.

### CODE REVIEW

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |

commit: https://github.com/percona/percona-xtradb-cluster-operator/pull/800/commits/dba174d9196163501c4f61c7d5c144f419e3ae7e
image: `perconalab/percona-xtradb-cluster-operator:PR-800-dba174d9`

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |

commit: https://github.com/percona/percona-xtradb-cluster-operator/pull/800/commits/dba174d9196163501c4f61c7d5c144f419e3ae7e
image: `perconalab/percona-xtradb-cluster-operator:PR-800-dba174d9`

File: pkg/controller/pxcbackup/controller.go [hunk 1]
```diff
@@ -107,102 +129,195 @@ func (r *ReconcilePerconaXtraDBClusterBackup) Reconcile(request reconcile.Reques
...
+	cr.SetFinalizers(finalizers)
+
+	err = r.client.Update(context.TODO(), cr)
+	if err != nil {
+		log.Error(err, "failed to update finalizers for backup", "backup", cr.Name)
+	}
+}
+
+func (r *ReconcilePerconaXtraDBClusterBackup) removeBackup(bucket, backup string, s3cli *minio.Client) error {
+	objs := s3cli.ListObjects(context.Background(), bucket,
```
Comment (delgod) [inline pkg/controller/pxcbackup/controller.go:300]: can we remove `.md5` file before start deleting all other files?
it is ok if md5 file not exists, but it is not ok if we cannot (don't have permissions) delete it.

**Review by delgod — CHANGES_REQUESTED** [review]

File: pkg/controller/pxcbackup/controller.go (hunk 1 above)
Comment (blez) [reply pkg/controller/pxcbackup/controller.go:303]: done

**Review by delgod — APPROVED** [review]

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| storage | passed |
| recreate | passed |
| haproxy | passed |
| upgrade-haproxy | passed |
| big-data | passed |
| scaling | passed |
| init-deploy | passed |
| scaling-proxysql | failed |
| self-healing | passed |
| restore-to-encrypted-cluster | passed |
| upgrade-proxysql | passed |
| limits | passed |
| self-healing-advanced | passed |
| security-context | failed |
| monitoring-2-0 | passed |
| demand-backup | passed |
| affinity | passed |
| smart-update | passed |
| operator-self-healing | passed |
| one-pod | passed |
| auto-tuning | passed |
| proxysql-sidecar-res-limits | passed |
| users | passed |
| upgrade-consistency | passed |
| tls-issue-self | passed |
| tls-issue-cert-manager | passed |
| tls-issue-cert-manager-ref | passed |
| validation-hook | passed |
| pitr | failed |

commit: https://github.com/percona/percona-xtradb-cluster-operator/pull/800/commits/8951ade63f6770a2d91cc29a7ba3fd1afaa4c94e
image: `perconalab/percona-xtradb-cluster-operator:PR-800-8951ade6`

File: pkg/controller/pxcbackup/controller.go [hunk 2]
```diff
@@ -107,102 +129,198 @@ func (r *ReconcilePerconaXtraDBClusterBackup) Reconcile(request reconcile.Reques
...
```
Comment (qJkee) [inline pkg/controller/pxcbackup/controller.go:303]: should we just call "delete" command, because we know the object name?

Comment (blez) [reply pkg/controller/pxcbackup/controller.go:303]: No, we know only prefix, not whole name

**Review by qJkee — APPROVED** [review]

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| upgrade-haproxy | passed |
| haproxy | passed |
| recreate | passed |
| scaling | passed |
| big-data | passed |
| storage | passed |
| init-deploy | passed |
| self-healing | passed |
| scaling-proxysql | passed |
| restore-to-encrypted-cluster | passed |
| upgrade-proxysql | passed |
| limits | passed |
| self-healing-advanced | passed |
| security-context | failed |
| monitoring-2-0 | passed |
| demand-backup | passed |
| affinity | passed |
| operator-self-healing | passed |
| one-pod | passed |
| auto-tuning | passed |
| proxysql-sidecar-res-limits | passed |
| users | passed |
| pitr | passed |
| tls-issue-self | passed |
| demand-backup-encrypted-with-tls | passed |
| tls-issue-cert-manager | passed |
| tls-issue-cert-manager-ref | passed |
| validation-hook | passed |
| scheduled-backup | failed |

commit: https://github.com/percona/percona-xtradb-cluster-operator/pull/800/commits/d6d6cf14a5388032ea8e0813795c3e4ca3fd2717
image: `perconalab/percona-xtradb-cluster-operator:PR-800-d6d6cf14`

**Review by delgod — APPROVED** [review]

**Pull request merged**